pkg/controllers/rollout/controller.go

```diff
@@ -478,12 +515,36 @@ func (r *Reconciler) pickBindingsToRoll(ctx context.Context, allBindings []*flee
 	if boundingCandidatesUnselectedIndex < len(boundingCandidates) {
 		staleUnselectedBinding = append(staleUnselectedBinding, boundingCandidates[boundingCandidatesUnselectedIndex:]...)
 	}
+	return toBeUpdatedBindingList, staleUnselectedBinding
+}
 
-	return toBeUpdatedBindingList, staleUnselectedBinding, true, nil
+func calculateMaxToRemove(crp *fleetv1beta1.ClusterResourcePlacement, targetNumber int, readyBindings, canBeUnavailableBindings []*fleetv1beta1.ClusterResourceBinding) int {
+	maxUnavailableNumber, _ := intstr.GetScaledValueFromIntOrPercent(crp.Spec.Strategy.RollingUpdate.MaxUnavailable, targetNumber, true)
+	minAvailableNumber := targetNumber - maxUnavailableNumber
+	// This is the lower bound of the number of bindings that can be available during the rolling update
+	// Since we can't predict the number of bindings that can be unavailable after they are applied, we don't take them into account
+	lowerBoundAvailableNumber := len(readyBindings) - len(canBeUnavailableBindings)
+	maxNumberToRemove := lowerBoundAvailableNumber - minAvailableNumber
+	klog.V(2).InfoS("Calculated the max number of bindings to remove", "clusterResourcePlacement", klog.KObj(crp),
+		"maxUnavailableNumber", maxUnavailableNumber, "minAvailableNumber", minAvailableNumber,
+		"lowerBoundAvailableBindings", lowerBoundAvailableNumber, "maxNumberOfBindingsToRemove", maxNumberToRemove)
+	return maxNumberToRemove
 }
 
+func calculateMaxToAdd(crp *fleetv1beta1.ClusterResourcePlacement, targetNumber int, canBeReadyBindings []*fleetv1beta1.ClusterResourceBinding) int {
+	maxSurgeNumber, _ := intstr.GetScaledValueFromIntOrPercent(crp.Spec.Strategy.RollingUpdate.MaxSurge, targetNumber, true)
+	maxReadyNumber := targetNumber + maxSurgeNumber
+	// This is the upper bound of the number of bindings that can be ready during the rolling update
+	// We count anything that still has work object on the hub cluster as can be ready since the member agent may have connection issue with the hub cluster
+	upperBoundReadyNumber := len(canBeReadyBindings)
+	maxNumberToAdd := maxReadyNumber - upperBoundReadyNumber
+
+	klog.V(2).InfoS("Calculated the max number of bindings to add", "clusterResourcePlacement", klog.KObj(crp),
+		"maxSurgeNumber", maxSurgeNumber, "maxReadyNumber", maxReadyNumber, "upperBoundReadyBindings",
+		upperBoundReadyNumber, "maxNumberOfBindingsToAdd", maxNumberToAdd)
+	return maxNumberToAdd
+}
```

Comment:
I am not sure if there is benefit of this type of refactor if we don't add UT for those functions.

Reply:
and looking at the logic, they seem to be too straightforward to warrantee UT

Reply:
I took them out of the function because of this error https://github.com/Azure/fleet/actions/runs/10154135990/job/28078753503. Most of the "refactoring" was just breaking up the function into smaller, more manageable functions.